# Name: Sunil Pradhan

## Project: 1
## Project Name: Mercedes-Benz Greener Manufacturing

In [1]:
#import libraries

import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import mean_squared_error
from math import sqrt

In [2]:
#read the test and train dataset
df_train=pd.read_csv("train.csv")
df_test=pd.read_csv("test.csv")

In [3]:
#printing the train dataset

df_train.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [4]:
#printing the test dataset

df_test.head()

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   
3   4  az  l   n  f  d  z  l  n    0  ...     0     0     0     1     0     0   
4   5   w  s  as  c  d  y  i  m    0  ...     1     0     0     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 377 columns]

In [5]:
df_train.columns

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=378)

In [6]:
df_test.columns

Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=377)

Dropping the column ID as it not required for model prediction

In [7]:
df_train=df_train.drop(['ID'], axis=1)
df_test=df_test.drop(['ID'], axis=1)

In [8]:
df_train.head()

y  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  \
0  130.81   k  v  at  a  d  u  j  o    0  ...     0     0     1     0     0   
1   88.53   k  t  av  e  d  y  l  o    0  ...     1     0     0     0     0   
2   76.26  az  w   n  c  d  x  j  x    0  ...     0     0     0     0     0   
3   80.62  az  t   n  f  d  x  l  e    0  ...     0     0     0     0     0   
4   78.02  az  v   n  f  d  h  d  n    0  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 377 columns]

In [9]:
df_test.head()

X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  ...  X375  X376  X377  X378  X379  \
0  az  v   n  f  d  t  a  w    0    0  ...     0     0     0     1     0   
1   t  b  ai  a  d  b  g  y    0    0  ...     0     0     1     0     0   
2  az  v  as  f  d  a  j  j    0    0  ...     0     0     0     1     0   
3  az  l   n  f  d  z  l  n    0    0  ...     0     0     0     1     0   
4   w  s  as  c  d  y  i  m    0    0  ...     1     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     0     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 376 columns]

### If for any column(s), the variance is equal to zero, then you need to remove those variable(s)

In [10]:
var_col_zero=df_train.var()[df_train.var()==0].index.values


<ipython-input-10-37cc858c5963>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  var_col_zero=df_train.var()[df_train.var()==0].index.values


In [11]:
#column index having zero variance
var_col_zero

array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

In [12]:
#dropping the column having zero variance
df_train1=df_train.drop(var_col_zero, axis=1)
df_test1=df_test.drop(var_col_zero, axis=1)

In [13]:
#checking the initial columns and final columns after droping column having variance=0
df_train.shape[1], df_train1.shape[1]

(377, 365)

In [14]:
df_test.shape[1], df_test1.shape[1]

(376, 364)

### Check for null and unique values for test and train sets

In [15]:
#checking for null values
np.sum(df_train1.isnull().sum())

0

In [16]:
np.sum(df_test1.isnull().sum())

0

In both train and test data set there is no null values

In [17]:
#checking for unique value
np.sum(df_train1.nunique().sum())

3452

In [18]:
np.sum(df_test1.nunique().sum())

908

There is 3452 unique values are in the train dataset and 908 unique values are in the dataset

In [19]:
#printing all the unique values of each column from train dataset

for i in df_train1:
    print("\nUnique value for the column",i)
    print(df_train1[i].unique())


Unique value for the column y
[130.81  88.53  76.26 ...  85.71 108.77  87.48]

Unique value for the column X0
['k' 'az' 't' 'al' 'o' 'w' 'j' 'h' 's' 'n' 'ay' 'f' 'x' 'y' 'aj' 'ak' 'am'
 'z' 'q' 'at' 'ap' 'v' 'af' 'a' 'e' 'ai' 'd' 'aq' 'c' 'aa' 'ba' 'as' 'i'
 'r' 'b' 'ax' 'bc' 'u' 'ad' 'au' 'm' 'l' 'aw' 'ao' 'ac' 'g' 'ab']

Unique value for the column X1
['v' 't' 'w' 'b' 'r' 'l' 's' 'aa' 'c' 'a' 'e' 'h' 'z' 'j' 'o' 'u' 'p' 'n'
 'i' 'y' 'd' 'f' 'm' 'k' 'g' 'q' 'ab']

Unique value for the column X2
['at' 'av' 'n' 'e' 'as' 'aq' 'r' 'ai' 'ak' 'm' 'a' 'k' 'ae' 's' 'f' 'd'
 'ag' 'ay' 'ac' 'ap' 'g' 'i' 'aw' 'y' 'b' 'ao' 'al' 'h' 'x' 'au' 't' 'an'
 'z' 'ah' 'p' 'am' 'j' 'q' 'af' 'l' 'aa' 'c' 'o' 'ar']

Unique value for the column X3
['a' 'e' 'c' 'f' 'd' 'b' 'g']

Unique value for the column X4
['d' 'b' 'c' 'a']

Unique value for the column X5
['u' 'y' 'x' 'h' 'g' 'f' 'j' 'i' 'd' 'c' 'af' 'ag' 'ab' 'ac' 'ad' 'ae'
 'ah' 'l' 'k' 'n' 'm' 'p' 'q' 's' 'r' 'v' 'w' 'o' 'aa']

Unique value for the colu

In [20]:
#printing all the unique values of each column from test dataset

for i in df_test1:
    print("\nUnique value for the column",i)
    print(df_train1[i].unique())


Unique value for the column X0
['k' 'az' 't' 'al' 'o' 'w' 'j' 'h' 's' 'n' 'ay' 'f' 'x' 'y' 'aj' 'ak' 'am'
 'z' 'q' 'at' 'ap' 'v' 'af' 'a' 'e' 'ai' 'd' 'aq' 'c' 'aa' 'ba' 'as' 'i'
 'r' 'b' 'ax' 'bc' 'u' 'ad' 'au' 'm' 'l' 'aw' 'ao' 'ac' 'g' 'ab']

Unique value for the column X1
['v' 't' 'w' 'b' 'r' 'l' 's' 'aa' 'c' 'a' 'e' 'h' 'z' 'j' 'o' 'u' 'p' 'n'
 'i' 'y' 'd' 'f' 'm' 'k' 'g' 'q' 'ab']

Unique value for the column X2
['at' 'av' 'n' 'e' 'as' 'aq' 'r' 'ai' 'ak' 'm' 'a' 'k' 'ae' 's' 'f' 'd'
 'ag' 'ay' 'ac' 'ap' 'g' 'i' 'aw' 'y' 'b' 'ao' 'al' 'h' 'x' 'au' 't' 'an'
 'z' 'ah' 'p' 'am' 'j' 'q' 'af' 'l' 'aa' 'c' 'o' 'ar']

Unique value for the column X3
['a' 'e' 'c' 'f' 'd' 'b' 'g']

Unique value for the column X4
['d' 'b' 'c' 'a']

Unique value for the column X5
['u' 'y' 'x' 'h' 'g' 'f' 'j' 'i' 'd' 'c' 'af' 'ag' 'ab' 'ac' 'ad' 'ae'
 'ah' 'l' 'k' 'n' 'm' 'p' 'q' 's' 'r' 'v' 'w' 'o' 'aa']

Unique value for the column X6
['j' 'l' 'd' 'h' 'i' 'a' 'g' 'c' 'k' 'e' 'f' 'b']

Unique value for the c

### Applying Label Encoder

In [21]:
#finding the column having datatype as object
obj_col=df_train1.describe(include=[object]).columns.values
obj_col

array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

In [22]:
#label encoder
le=LabelEncoder()
for i in obj_col:
    le.fit(df_train1[i].append(df_test1[i]).values)
    df_train1[i]=le.transform(df_train1[i])
    df_test1[i]=le.transform(df_test1[i])

In [23]:
#create X and y
X=df_train1.drop(['y'], axis=1)
y=df_train1['y']

In [24]:
# create test and train split
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest=train_test_split(X,y,test_size=0.3,random_state=7)

### Performing dimensionality reduction

In [25]:
from sklearn.decomposition import PCA

In [26]:
#using PCA decomposition of the data from 364 components with 98% variance to lower dimension of 12 compnonents
pca=PCA(0.98, svd_solver='full')
sklearn_pca=pca.fit(X)

In [27]:
sklearn_pca.n_components_

12

In [28]:
sklearn_pca.explained_variance_ratio_

array([0.40868988, 0.21758508, 0.13120081, 0.10783522, 0.08165248,
       0.0140934 , 0.00660951, 0.00384659, 0.00260289, 0.00214378,
       0.00209857, 0.00180388])

### predicting using XGBoost

In [29]:
pca_xtrain=pd.DataFrame(pca.transform(xtrain))
pca_xtest=pd.DataFrame(pca.transform(xtest))
pca_df_test1=pd.DataFrame(pca.transform(df_test1))

In [30]:
import xgboost as xgb

In [31]:
xgb_reg=xgb.XGBRegressor(objective='reg:squarederror',learning_rate=0.1)
xgb_reg.fit(pca_xtrain,ytrain)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
             gamma=0, gpu_id=-1, importance_type=None,
             interaction_constraints='', learning_rate=0.1, max_delta_step=0,
             max_depth=6, min_child_weight=1, missing=nan,
             monotone_constraints='()', n_estimators=100, n_jobs=8,
             num_parallel_tree=1, predictor='auto', random_state=0, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method='exact',
             validate_parameters=1, verbosity=None)

In [32]:
xgb_reg_pred_ytest=xgb_reg.predict(pca_xtest)

In [33]:
print(sqrt(mean_squared_error(ytest,xgb_reg_pred_ytest)))

10.47229744101521


In [34]:
xgb_RFreg=xgb.XGBRFRegressor(objective='reg:squarederror',learning_rate=1)
xgb_RFreg.fit(pca_xtrain,ytrain)

XGBRFRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
               colsample_bytree=1, enable_categorical=False, gamma=0, gpu_id=-1,
               importance_type=None, interaction_constraints='',
               learning_rate=1, max_delta_step=0, max_depth=6,
               min_child_weight=1, missing=nan, monotone_constraints='()',
               n_estimators=100, n_jobs=8, num_parallel_tree=100,
               objective='reg:squarederror', predictor='auto', random_state=0,
               reg_alpha=0, scale_pos_weight=1, tree_method='exact',
               validate_parameters=1, verbosity=None)

In [35]:
xgb_RFreg_pred_ytest=xgb_RFreg.predict(pca_xtest)

In [36]:
print(sqrt(mean_squared_error(ytest,xgb_RFreg_pred_ytest)))

10.64404619286274
